src/cascadia/TerminalApp/Pane.cpp

```diff
@@ -187,6 +187,95 @@ bool Pane::ResizePane(const Direction& direction)
+
+    const auto newlyFocusedChild = focusSecond ? _secondChild : _firstChild;
+
+    // If the child we want to move focus to is _already_ focused, return false,
+    // to try and let our parent figure it out.
+    if (newlyFocusedChild->WasLastFocused())
+    {
+        return false;
+    }
+
+    // Transfer focus to our child, and update the focus of our tree.
+    newlyFocusedChild->_FocusFirstChild();
+    UpdateFocus();
+
+    return true;
+}
+
+// Method Description:
+// - Attempts to move focus to one of our children. If we have a focused child,
+//   we'll try to move the focus in the direction requested.
+//   - If there isn't a pane that exists as a child of this pane in the correct
+//     direction, we'll return false. This will indicate to our parent that they
+//     should try and move the focus themselves. In this way, the focus can move
+//     up and down the tree to the correct pane.
+// - This method is _very_ similar to ResizePane. Both are trying to fine the
+//   right separator to move (focus) in a direction.
+// Arguments:
+// - direction: The direction to move the focus in.
+// Return Value:
+// - true if we or a child handled this focus move request.
+bool Pane::NavigateFocus(const Direction& direction)
+{
+    // If we're a leaf, do nothing. We can't possibly have a descendant with a
+    // separator the correct direction.
+    if (_IsLeaf())
+    {
+        return false;
+    }
+
+    // Check if either our first or second child is the currently focused leaf.
+    // If it is, and the requested move direction matches our separator, then
+    // we're the pane that needs to handle this focus move.
+    const bool firstIsFocused = _firstChild->_IsLeaf() && _firstChild->_lastFocused;
+    const bool secondIsFocused = _secondChild->_IsLeaf() && _secondChild->_lastFocused;
+    if (firstIsFocused || secondIsFocused)
+    {
+        return _NavigateFocus(direction);
+    }
+    else
+    {
+        // If neither of our children were the focused leaf, then recurse into
+        // our children and see if they can handle the focus move.
+        // For each child, if it has a focused descendant, try having that child
+        // handle the focus move.
+        // If the child wasn't able to handle the focus move, it's possible that
+        // there were no descendants with a separator the correct direction. If
+        // our separator _is_ the correct direction, then we should be the pane
+        // to move focus into our other child. Otherwise, just return false, as
+        // we couldn't handle it either.
+        if ((!_firstChild->_IsLeaf()) && _firstChild->_HasFocusedChild())
+        {
+            return _firstChild->NavigateFocus(direction) || _NavigateFocus(direction);
+        }
+        else if ((!_secondChild->_IsLeaf()) && _secondChild->_HasFocusedChild())
+        {
+            return _secondChild->NavigateFocus(direction) || _NavigateFocus(direction);
+        }
+    }
+    return false;

```

Comment:
I'm worried that there's a `_NavigateFocus` in three of the .. four? states for this conditional. Is there a better way to structure this code?

Reply:
(we discussed on chat that this is fine the way it is an there's not really a better structure for it)